### DISCUSSION TITLE
How to stop the job if it took more than the specified IntervalTrigger?

### DISCUSSION THREAD

**Discussion by BabakAmini** [agronholm/apscheduler#778] [Q&A]
I'm using apscheduler v4.0.0a2 to retrieve data from an API every second. The request takes approximately 200 ms, and I need an extra 200 ms for processing the data, so this works as expected most of the time. However, in some situations, the total time of request and processing takes more than one second. By incorporating a timeout value, I can regulate the duration of the request but, due to the blocking nature of the processing task, I lack direct influence over its completion time. In such cases, if the CPU, for instance, is occupied, the processing time might exceed the expected duration.

So my question is, how can I cancel that blocking task if the time interval is triggered? In the latest version of apscheduler, it was possible to achieve this functionality by including the `max_running_jobs` keyword argument when using the `add_schedule()` method, as explained [here](https://apscheduler.readthedocs.io/en/master/userguide.html#limiting-the-number-of-concurrently-executing-instances-of-a-job). However, this feature is no longer available in version 4.0.0a2.

**Comment by agronholm** [comment]
There is an open PR to add that option back, but that's not the issue. If you have a function running in a thread, how do you propose to stop it from the outside?

**EDIT**: I forgot I already merged it, so it will be in the next alpha.

**Reply by BabakAmini** [reply]
Unfortunately, it looks like I misunderstood about `max_running_jobs`. I actually want to terminate the task if the total value of running exceeds one second. That argument forces the task to be complete even if it takes more than one second and doesn't fire the next event until it returns.

My code is already complex enough to force me to avoid using other methods, such as threading and signaling, which I'm new to. I currently have a well-structured schedule for my tasks using `apscheduler`. Each task has its own set of scheduling parameters and behaviors. The tasks also depend on the processing results of each other. Because of this, I find it preferable to continue using `apscheduler` and let it handle any exceptions that may arise. Do you have any idea in this regard?

**Reply by BabakAmini** [reply]
Hello @agronholm  , I'm still looking for a workaround for my issue. I upgraded to 4.0.0a3 and now `max_running_jobs` is added to `add_schedule`. It works perfectly if I want to complete a currently running job even when a new job arrives. Since my task has a specified time limit (in this case, one second), it is essential for the running job to be terminated when a new job begins. I figured out there is a `coalesce` parameter that seems to be what I need, but I have no clue how to use it.

Any suggestions would be highly appreciated.

**Reply by agronholm** [reply]
The scheduler still doesn't enforce the setting, and there is no way yet to actually cancel any jobs. Therefore the target function would have to enforce the time limit on itself for the time being.
